### ISSUE TITLE
Reset password broken for accounts with upper case email address.

### ISSUE THREAD

**Issue by david-uhlig** [RocketChat/Rocket.Chat#31518]
### Description:

Users with upper case characters in their e-mail address, e.g. John.Doe@example.com are not recognized when requesting a password reset via the web frontend.

### Steps to reproduce:

1. With admin priviledges, go to /admin/users/new
2. Create a new user with an e-mail address containing at least one upper case character, e.g. John.Doe@example.com
3. Logout
4. Go to /home, click on "Forgot your password?", enter the e-mail exactly as created, e.g. John.Doe@example.com
5. Sign in with admin priviledges and check the logs: /admin/reports

### Expected behavior:

The user should receive a password reset e-mail with instructions to reset their password.

### Actual behavior:

The user does not receive an e-mail, because the request fails.
Log entry:

```json
{"level":50,"time":"2024-01-23T13:50:38.369Z","pid":1,"hostname":"<redacted>","name":"System","err":
{"type":"errorClass","message":"No such email for user. [403]","stack":"Error: No such email for user. [403]<br>    at AccountsServer._handleError (packages/accounts-base/accounts_server.js:1488:19)<br>    at AccountsServer.Accounts.generateResetToken (packages/accounts-password/password_server.js:415:14)<br>    at AccountsServer.Accounts.sendResetPasswordEmail (packages/accounts-password/password_server.js:537:14)<br>    at server/methods/sendForgotPasswordEmail.ts:36:13<br>    at /app/bundle/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/fiber_pool.js:43:40","isClientSafe":true,"error":403,"reason":"No such email for user.","errorType":"Meteor.Error"}
,"msg":"No such email for user. [403]"}
```

### Server Setup Information:

- Version of Rocket.Chat Server: 6.5.2.
- Operating System: Ubuntu 22.04 LTS
- Deployment Method: Docker
- Number of Running Instances: 1
- DB Replicaset Oplog: 
- NodeJS Version: 
- MongoDB Version: 5.0

### Client Setup Information

- Desktop App or Browser Version: Firefox 
- Operating System: Windows 10

**Comment by 1999matrix** [comment]
Can you kindly share it's video , because i am getting reset password mail .

**Comment by 1999matrix** [comment]
@david-uhlig can you kindly share it's video , because i am getting reset password mail .

**Comment by hardikbhatia777** [comment]
Hey again @david-uhlig. Thanks for reporting the issue! I was able to reproduce this issue.

It seems like the email parsing system needs an overhaul. There are far too many edge cases that cause failure in some form or the other. A case could be made that all emails should pass through a normalization function which trims the whitespaces (as you mentioned in your last report) and converts the email to lower case and so on. 

What do you think? Should we disallow the usage of uppercase characters entirely, or perhaps convert the email to lowercase? Any user suggestions are highly appreciated!

My thoughts on this:
Since uppercase letter aren't really a thing in email addresses, it is highly unlikely that a user will deliberately be registered this way. This can, however, be a mistake on the administrator's side or the user's side while entering email address. (Note: Logging seems to not be case sensitive, however the password reset field is.) 

In my opinion, we should have an email normalization function which converts all emails into proper format, and modifying all existing email input fields to simply pass through this function upon being entered before they are used for further process / storage. This includes trimming whitespaces, converting everything to lowercase, and any other cases that could arise. This would be a more permanent solution imo. If a file like this already exists, it can be updated. If not, I can make one and convert everything to use it on the API level (along with some tests).

Let me know your thoughts on this @hugocostadev @dougfabris @david-uhlig.

**Comment by david-uhlig** [comment]
Mails should also always get sent to the mail address that is stored in the database. Not to an address provided in other forms, i.e. the password reset form, even if they are normalized. Otherwise, a security vulnerability is waiting to happen when john.doe@example.com can request a password reset for John.Doe@example.com.

See: https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/server/methods/sendForgotPasswordEmail.ts

```ts
Meteor.methods<ServerMethods>({
	async sendForgotPasswordEmail(to) {
		check(to, String);

		const email = to.trim().toLowerCase();

		const user = await Users.findOneByEmailAddress(email, { projection: { _id: 1 } });

		if (!user) {
			return true;
		}

		if (user.services && !user.services.password) {
			if (!settings.get('Accounts_AllowPasswordChangeForOAuthUsers')) {
				return false;
			}
		}

		try {
			Accounts.sendResetPasswordEmail(user._id, email);  // <----
			return true;
		} catch (error) {
			SystemLogger.error(error);
		}
	},
});
```

Not sure if I should open an additional bug report for this. Please advise.

**Comment by Yashraj7890** [comment]
Hey @hardikbhatia777 and @david-uhlig , after going through the conversation above, I think not allowing case sensitive emails to be registered will be a good solution as there is no possibility of a valid email to be case sensitive and I guess this approach may also not cause any other anomaly.

**Comment by david-uhlig** [comment]
> no possibility of a valid email to be case sensitive

That's not true, [the local part can be case-sensitive](https://stackoverflow.com/a/9808332) as described above.

**Comment by hardikbhatia777** [comment]
I was waiting on a reply from the maintainers first, although I must disagree @Yashraj7890. While it is _**generally**_ fine to assume that emails are case-insensitive as an end user, the actual software must be designed to match internationally accepted standards and regulations. 

Thanks for the additional tip @david-uhlig. It's best to proceed according to the decision which shall be made by the staff. For the time being, since this is related, I don't think this needs to be a separate issue. I'll keep this in mind for now.

**Comment by Yashraj7890** [comment]
Apologies for the slight mistake😅, but it is also mentioned in that post that " no widely used mail systems distinguish different addresses based on case" and I guess an account with a case sensitive email address would be extremely rare, even in discord every email is treated as case insensitive I tried logging in with my gmail address after changing the casing and it successfully logs in which means that there will be no two separate accounts distinguished on the basis of casing, so I guess it would be fine.

**Comment by hardikbhatia777** [comment]
> so I guess it would be fine.
> 

Your take is totally understandable, but code can't be pushed to production with the mentality of "I guess it would be fine". That's a recipe for disaster and makes the code less robust. 

Let's wait for the staff to reach a decision internally about this first :) 

Just for your reference, RFC 5321 2.4 is the regulation that prevents a 'convert it all to lowercase' solution.

**Comment by Yashraj7890** [comment]
Noted 😁👍.

**Comment by alihamza1221** [comment]
@david-uhlig I have created a PR to resolve this issue it was basically  due to the convertion of email to lowecase() while sending to Accout.Meteor() so, as it **does't match with the orignal** email it gives error the details are mentioned in the pr.
